lib/npg_qc/autoqc/checks/haplotag_metrics.pm

```diff
@@ -0,0 +1,180 @@
+
+package npg_qc::autoqc::checks::haplotag_metrics;
+
+use Moose;
+use MooseX::StrictConstructor;
+use namespace::autoclean;
+use Carp;
+use File::Basename;
+use File::Spec::Functions qw(catfile);
+use Perl6::Slurp;
+use Readonly;
+use st::api::lims;
+use IPC::Open3 'open3';
+local $SIG{CHLD} = 'IGNORE';
+
+extends qw( npg_qc::autoqc::checks::check );
+
+our $VERSION = '0';
+
+Readonly::Scalar our $PASS_RATE => 0.8;
+Readonly::Scalar our $FILE_EXT => q[SamHaplotag_Clear_BC];
+
+has '_file_path_root' => ( isa        => 'Str',
+                           is         => 'ro',
+                           lazy_build => 1,);
+
+sub _build__file_path_root {
+  my $self = shift;
+  my $path;
+  if ($self->has_filename_root && $self->has_qc_in) {
+    $path = catfile $self->qc_in, $self->filename_root;
+  } else {
+    ($path) = $self->input_files->[0] =~ /\A(.+)\..+\Z/smx;
+  }
+
+  return $path;
+}
+
+override 'can_run' => sub {
+  my $self = shift;
+  my $lims;
+  if ($self->has_rpt_list) {
+    $lims = st::api::lims->new(rpt_list => $self->rpt_list);
+  } else {
+    $lims = st::api::lims->new(id_run=>$self->id_run, position=>$self->position);
+  }
+  my $library_type = $lims->library_type;
+  my $is_haplotag_lib = $library_type && ($library_type =~ /Haplotagging/smx);
+  return $is_haplotag_lib;
+};
+
+
+override 'execute' => sub {
+  my $self = shift;
+
+  if (!$self->can_run()) {
+    return 1;
+  }
+
+  super();
+
+  my $clear_file = $self->_file_path_root . q[._SamHaplotag_Clear_BC];
+  my $count;
+  my ($writer, $reader, $err);
+  open3($writer, $reader, $err, "wc -l $clear_file");
+  $count = <$reader>;
+  my ($clear_count) = $count =~ /(\d+)/smx;
+  $clear_count ||= 1; $clear_count -= 1;
+
+  my $unclear_file = $self->_file_path_root . q[._SamHaplotag_UnClear_BC];
+  open3($writer, $reader, $err, "wc -l $unclear_file");
+  $count = <$reader>;
+  my ($unclear_count) = $count =~ /(\d+)/smx;
+  $unclear_count ||= 1; $unclear_count -= 1;
+
+  my $missing_file = $self->_file_path_root . q[._SamHaplotag_Missing_BC_QT_tags];
+  open3($writer, $reader, $err, "wc -l $missing_file");
+  $count = <$reader>;
+  my ($missing_count) = $count =~ /(\d+)/smx;
+  $missing_count ||= 1; $missing_count -= 1;
+
+  $self->result->{'clear_file'} = $clear_file;
+  $self->result->{'unclear_file'} = $unclear_file;
+  $self->result->{'missing_file'} = $missing_file;
+
+  $self->result->{'clear_count'} = $clear_count;
+  $self->result->{'unclear_count'} = $unclear_count;
+  $self->result->{'missing_count'} = $missing_count;
+
+  my $pass = ($clear_count + $unclear_count) / ($clear_count + $unclear_count + $missing_count);
```

Comment:
Is this always 100% for our data?

Reply:
The SamHaplotag program puts reads with no BX tag assigned into the 'Missing' file, and all others into the 'Clear' and 'Unclear' files.
We may want to change the $PASS_RATE from 0.8 to 0.9 (my notes (ie handwriting) is a little unclear on that.

Reply:
Does 'Missing' ever have any data for our pipelines though? I think all our data has BC auxtag to do the calculation on. I _suspect_ dud haplotags get counted in 'Unclear'.

N.b. @dozy a good example to check would be where we've done the "revcomp" incorrectly - that shoudl come out as a clear fail.

Reply:
I'm not clear what 'dud haplotags' means in this context. The test was supposed to be 'N% of reads assigned a BX tag'. We expect most of our data to pass this test, but wanted to check for an occasional fail.

Reply:
Are we always getting BX assigned? If it's always 100% then that criteria isn't useful - misapplying the revcomp produces data which breaks downstream analysis, as such it is a good 1st test whether what we're doing (the pass criteria) has value.